File: payola/web/server/public/javascripts/client/cz/payola/web/client/views/graph/EdgeView.js
goog.provide('cz.payola.web.client.views.graph.EdgeView');
goog.provide('cz.payola.web.client.views.graph.Quadrant');
goog.require('cz.payola.web.client.views.Constants');
goog.require('cz.payola.web.client.views.Point');
goog.require('cz.payola.web.client.views.graph.InformationView');
goog.require('cz.payola.web.client.views.graph.View');
goog.require('scala.math');
cz.payola.web.client.views.graph.EdgeView = function(edgeModel, originView, destinationView) {
var self = this;
self.edgeModel = edgeModel;
self.originView = originView;
self.destinationView = destinationView;
self.information = new cz.payola.web.client.views.graph.InformationView(self.edgeModel);
goog.base(self);};
goog.inherits(cz.payola.web.client.views.graph.EdgeView, cz.payola.web.client.views.graph.View);
cz.payola.web.client.views.graph.EdgeView.prototype.isSelected = function() {
var self = this;
return (self.originView.selected || self.destinationView.selected);
};
cz.payola.web.client.views.graph.EdgeView.prototype.draw = function(context, color, positionCorrection) {
var self = this;
var A = self.originView.position;
var B = self.destinationView.position;
var ctrl1 = new cz.payola.web.client.views.Point(0.0, 0.0);
var ctrl2 = new cz.payola.web.client.views.Point(0.0, 0.0);
var diff = new cz.payola.web.client.views.Point(scala.math.abs((A.x - B.x)), scala.math.abs((A.y - B.y)));
var quadrant = (function() {
if ((A.x <= B.x)) {
return (function() {
if ((A.y <= B.y)) {
return cz.payola.web.client.views.graph.Quadrant.RightBottom;
} else {
return cz.payola.web.client.views.graph.Quadrant.RightTop;
}})();
} else {
return (function() {
if ((A.y <= B.y)) {
return cz.payola.web.client.views.graph.Quadrant.LeftBottom;
} else {
return cz.payola.web.client.views.graph.Quadrant.LeftTop;
}})();
}})();
if ((diff.x >= diff.y)) {
(function($selector$1) {
if (($selector$1 === cz.payola.web.client.views.graph.Quadrant.RightBottom) || ($selector$1 === cz.payola.web.client.views.graph.Quadrant.RightTop)) {
ctrl1.x = (A.x + (diff.x / cz.payola.web.client.views.Constants.EdgeSIndex));
ctrl1.y = A.y;
ctrl2.x = (B.x - (diff.x / cz.payola.web.client.views.Constants.EdgeSIndex));
ctrl2.y = B.y;

return;
}
if (($selector$1 === cz.payola.web.client.views.graph.Quadrant.LeftBottom) || ($selector$1 === cz.payola.web.client.views.graph.Quadrant.LeftTop)) {
ctrl1.x = (A.x - (diff.x / cz.payola.web.client.views.Constants.EdgeSIndex));
ctrl1.y = A.y;
ctrl2.x = (B.x + (diff.x / cz.payola.web.client.views.Constants.EdgeSIndex));
ctrl2.y = B.y;

return;
}
})(quadrant);
} else {
(function($selector$2) {
if (($selector$2 === cz.payola.web.client.views.graph.Quadrant.RightBottom) || ($selector$2 === cz.payola.web.client.views.graph.Quadrant.LeftBottom)) {
ctrl1.x = A.x;
ctrl1.y = (A.y + (diff.y / cz.payola.web.client.views.Constants.EdgeSIndex));
ctrl2.x = B.x;
ctrl2.y = (B.y - (diff.y / cz.payola.web.client.views.Constants.EdgeSIndex));

return;
}
if (($selector$2 === cz.payola.web.client.views.graph.Quadrant.RightTop) || ($selector$2 === cz.payola.web.client.views.graph.Quadrant.LeftTop)) {
ctrl1.x = A.x;
ctrl1.y = (A.y - (diff.y / cz.payola.web.client.views.Constants.EdgeSIndex));
ctrl2.x = B.x;
ctrl2.y = (B.y + (diff.y / cz.payola.web.client.views.Constants.EdgeSIndex));

return;
}
})(quadrant);
}
var colorToUse = (function() {
if ((color != null)) {
return color;
} else {
return (function() {
if (self.isSelected()) {
return cz.payola.web.client.views.Constants.ColorEdgeSelect;
} else {
return cz.payola.web.client.views.Constants.ColorEdge;
}})();
}})();
var correction = (function() {
if ((positionCorrection != null)) {
return positionCorrection.toVector();
} else {
return new cz.payola.web.client.views.Point(0.0, 0.0).toVector();
}})();
self.drawBezierCurve(context, ctrl1.$plus(correction), ctrl2.$plus(correction), A.$plus(correction), B.$plus(correction), cz.payola.web.client.views.Constants.EdgeWidth, colorToUse);

};
cz.payola.web.client.views.graph.EdgeView.prototype.metaClass_ = new s2js.MetaClass('cz.payola.web.client.views.graph.EdgeView', [cz.payola.web.client.views.graph.View]);
cz.payola.web.client.views.graph.Quadrant.RightBottom = 1;
cz.payola.web.client.views.graph.Quadrant.LeftBottom = 2;
cz.payola.web.client.views.graph.Quadrant.LeftTop = 3;
cz.payola.web.client.views.graph.Quadrant.RightTop = 4;
cz.payola.web.client.views.graph.Quadrant.metaClass_ = new s2js.MetaClass('cz.payola.web.client.views.graph.Quadrant', []);

File: payola/web/client/src/main/scala/cz/payola/web/client/views/plugins/visual/graph/EdgeView.scala
package cz.payola.web.client.views.plugins.visual.graph

import cz.payola.common.rdf.Edge
import s2js.adapters.js.dom.CanvasRenderingContext2D
import cz.payola.web.client.views.plugins.visual._
import settings.{TextSettingsModel, EdgeSettingsModel}

/**
  * Structure used during draw function of EdgeView. Helps to indicate position of vertices to each other.
  */
private object Quadrant
{
    //TODO correct to real enum
    val RightBottom = 1

    val LeftBottom = 2

    val LeftTop = 3

    val RightTop = 4
}

/**
  * Graphical representation of Edge object in the drawn graph.
  * @param edgeModel the object graphically represented by this class
  * @param originView the vertex object representing origin of this edge
  * @param destinationView of this graphical representation in drawing space
  */
class EdgeView(val edgeModel: Edge, val originView: VertexView, val destinationView: VertexView,
    val settings: EdgeSettingsModel, settingsText: TextSettingsModel) extends View {
    
    /**
      * Textual data that should be visualised with this edge ("over this edge").
      */
    val information: InformationView = new InformationView(edgeModel, settingsText)

    /**
      * Indicator of selection of this graphs element. Is used during color selection in draw function.
      * @return true if one of the edges vertices is selected.
      */
    def isSelected: Boolean = {
        originView.selected || destinationView.selected
    }

    /**
      * Indicator of selection of this graphs element. Is not used by inner mechanics.
      * @return true if both edges vertices are selected.
      */
    def areBothVerticesSelected: Boolean = {
        originView.selected && destinationView.selected
    }

    private def prepareBezierCurve(context: CanvasRenderingContext2D, color: Color, correction: Vector) {

        val A = originView.position
        val B = destinationView.position

        val ctrl1 = Point(0, 0)
        val ctrl2 = Point(0, 0)

        val diff = Point(scala.math.abs(A.x - B.x), scala.math.abs(A.y - B.y))

        //quadrant of coordinate system
        val quadrant = {
            //quadrant of destination
            if (A.x <= B.x) {
                if (A.y <= B.y) {
                    Quadrant.RightBottom
                } else {
                    Quadrant.RightTop
                }
            } else {
                if (A.y <= B.y) {
                    Quadrant.LeftBottom
                } else {
                    Quadrant.LeftTop
                }
            }
        }

        if (diff.x >= diff.y) {
            //connecting left/right sides of vertices
            quadrant match {
                case Quadrant.RightBottom | Quadrant.RightTop =>
                    //we are in (0, pi/4] or in (pi7/4, 2pi]
                    ctrl1.x = A.x + diff.x / settings.straightenIndex
                    ctrl1.y = A.y
                    ctrl2.x = B.x - diff.x / settings.straightenIndex
                    ctrl2.y = B.y
                case Quadrant.LeftBottom | Quadrant.LeftTop =>
                    //we are in (pi3/4, pi] or in (pi, pi5/4]
                    ctrl1.x = A.x - diff.x / settings.straightenIndex
                    ctrl1.y = A.y
                    ctrl2.x = B.x + diff.x / settings.straightenIndex
                    ctrl2.y = B.y
            }
        } else {
            //connecting top/bottom sides of vertices
            quadrant match {
                case Quadrant.RightBottom | Quadrant.LeftBottom =>
                    //we are in (pi/4, pi/2] or in (pi/2, pi3/4]
                    ctrl1.x = A.x
                    ctrl1.y = A.y + diff.y / settings.straightenIndex
                    ctrl2.x = B.x
                    ctrl2.y = B.y - diff.y / settings.straightenIndex
                case Quadrant.RightTop | Quadrant.LeftTop =>
                    //we are in (pi5/4, pi3/2] or in (pi3/2, pi7/4]
                    ctrl1.x = A.x
                    ctrl1.y = A.y - diff.y / settings.straightenIndex
                    ctrl2.x = B.x
                    ctrl2.y = B.y + diff.y / settings.straightenIndex
            }
        }
        
        drawBezierCurve(context, ctrl1 + correction, ctrl2 + correction, A + correction, B + correction,
            settings.width, color)
    }
    
    private def prepareStraight(context: CanvasRenderingContext2D, color: Color, correction: Vector) {
        
        drawStraightLine(context,
            destinationView.position + correction, originView.position + correction,
            settings.width, color)
    }
    
    def draw(context: CanvasRenderingContext2D, color: Option[Color], positionCorrection: Vector) {

        drawQuick(context, color, positionCorrection)
    }

    def drawQuick(context: CanvasRenderingContext2D, color: Option[Color], positionCorrection: Vector) {

        val colorToUse = color.getOrElse(
            if(isSelected) settings.colorSelected else settings.color
        )

        if(1 <= settings.straightenIndex && settings.straightenIndex <= 6) {
            prepareBezierCurve(context, colorToUse, Vector.Zero)
        } else {
            prepareStraight(context, colorToUse, Vector.Zero)
        }
    }

    override def toString: String = {
        "["+originView.toString+"-"+edgeModel.toString+"-"+destinationView.toString+"]"
    }
}
